### ISSUE TITLE
Apollo client requires react ?!?

### ISSUE THREAD

**Issue by Stradivario** [apollographql/apollo-client#8958]
Hi guys!

Thank you for the great project it opens up  many possibilities using graphql!

I am an early bird user from the moment that apollo introduced apollo-client!

I have created a package called `@rxdi/graphql-client ` which helps me to wrap all of the `apollo-client logic and initialize it inside a dependency injection called @rxdi.

So far so good but i decided to upgrade `apollo-client` to a new `3.*.*` version using `@apollo/client` package and i got a bundle error that says `Cannot resolve dependency 'react'`

```
  1 | import { invariant } from "../../utilities/globals/index.js";
> 2 | import * as React from 'react';
    |                       ^
  3 | import { getApolloContext } from "./ApolloContext.js";
  4 | export var ApolloProvider = function (_a) {
  5 |     var client = _a.client, children = _a.children;
```

So i have investigated a little bit and i found out that inside the barrel export ['index.ts'](https://github.com/apollographql/apollo-client/blob/main/src/index.ts) there is a line `export * from './react';` i was shocked to see this so i follow it up and i found this file https://github.com/apollographql/apollo-client/blob/main/src/react/context/ApolloProvider.tsx

1. So i am confused why react doesn't present as a dependency but it is required inside the code ?
2. Why even there is a code related to a `framework` which tight couples the `@apollo/client` to `react` ?
3. I now need to fork and remove all of this code in order to get the new version of `@apollo/client` does that sounds right ?
4. Does now `@apollo/client` can be called as `react` framework for graphql so i can skip next time when i want to install something which was framework agnostic ?

I am not a React user i have my own framework build using Web Components if you want you can check it out  using this starter https://github.com/rxdi/starter-client-side-lit-html

By the way this starter is using also `apollo-client` for interacting with Elon Musk Space X API

One way to mitigate this issue is to try to setup the `@apollo/client` not using the barrel export but instead a specific one by path. I have found that `@apollo/client/core` suits me well so wish me luck ! :)

Please don't get me wrong i like every framework but i decided to invest time and build my own stack and i am relying to a framework agnostic packages.

Cheers,
Kristiyan Tachev

**Comment by jpvajda** [comment]
This could be an interesting consideration for AC 4.0, So i'll add it to our milestone for consideration. Cc @bignimbus

**Comment by milesj** [comment]
Just ran into this while trying to use Apollo with Solid.

**Comment by gengns** [comment]
Same here when using Apollo with Svelte.

**Comment by andreasonny83** [comment]
It is very annoying to have a peerDependency to React. Importing from `@apollo/client/core` does not really solve the problem since that module is still living under the `@apollo/client` package which has a React dependency.
The correct way to solve this problem will be to separate the `@apollo/client/core` under a dedicated `@apollo/client-core` library React agnostic.

**Comment by phryneas** [comment]
@andreasonny83 `react` is only an [optional peer dependency](https://github.com/apollographql/apollo-client/blob/0aecf3e6fbe4fb371f4596305dc2a1cf82022f14/package.json#L77C1-L82C7) - your bundler should not force you to install it, only install it correctly in your `node_modules` folder structure in a way that is discoverable from Apollo Client if you already have `react` in your dependencies.

**Comment by andreasonny83** [comment]
@phryneas that is true, however the `peerDependenciesMeta` was only introduced by NPM v7 meaning that using older versions on NPM will result on an error when trying to consume a library with a dependency on Apollo Client.
Can I ask what is the reason for Apollo to keep the Core package under `@apollo/client` rather than exporting it to a separate npm package?

**Comment by phryneas** [comment]
@andreasonny83 npm v7 shipped with node 15, which went End of Life 2 years ago. Node 16 (the oldest officially supported node version, which will receive support for one more month at this point) already ships with npm v8.  
In one month, the oldest officially supported node version still receiving security updates will be node 18, which ships with npm 9.5.  

> Can I ask what is the reason for Apollo to keep the Core package under @apollo/client rather than exporting it to a separate npm package?

I'm sorry, but that decision predates me. 
I would guess it was mainly to help users keep their packages up-to-date, since before `@apollo/client` you had to install about 10 different packages and had to keep all of them in sync, which was very cumbersome and lead to a lot of bugs and frustration.

**Comment by jerelmiller** [comment]
Hey all, thanks for your patience! This work has been completed in #12300 and will be part of the 4.0 release (slated for later this year). All React-related exports have been moved to `@apollo/client/react`. React is no longer be required to import from `@apollo/client`.

We'll begin releasing alphas in the near future if you're looking to test this out. If you're interested in trying this out, be on the lookout for announcements related to 4.0. Thanks!
